proxy/ProxyClientSession.cc

```diff
@@ -124,43 +118,35 @@ ProxyClientSession::state_api_callout(int event, void *data)
   case EVENT_NONE:
   case EVENT_INTERVAL:
   case TS_EVENT_HTTP_CONTINUE:
-    if (likely(is_valid_hook(this->api_hookid))) {
-      if (this->api_current == nullptr && this->api_scope == API_HOOK_SCOPE_GLOBAL) {
-        this->api_current = http_global_hooks->get(this->api_hookid);
-        this->api_scope   = API_HOOK_SCOPE_LOCAL;
-      }
-
-      if (this->api_current == nullptr && this->api_scope == API_HOOK_SCOPE_LOCAL) {
-        this->api_current = ssn_hook_get(this->api_hookid);
-        this->api_scope   = API_HOOK_SCOPE_NONE;
-      }
+    if (nullptr == cur_hook) {
+      /// Get the next hook to invoke from HttpHookState
+      cur_hook = hook_state.getNext();
+    }
+    if (nullptr != cur_hook) {
+      bool plugin_lock    = false;
+      APIHook const *hook = cur_hook;
 
-      if (this->api_current) {
-        bool plugin_lock = false;
-        APIHook *hook    = this->api_current;
-        Ptr<ProxyMutex> plugin_mutex;
-
-        if (hook->m_cont->mutex) {
-          plugin_mutex = hook->m_cont->mutex;
-          plugin_lock  = MUTEX_TAKE_TRY_LOCK(hook->m_cont->mutex, mutex->thread_holding);
-          if (!plugin_lock) {
-            SET_HANDLER(&ProxyClientSession::state_api_callout);
-            if (!schedule_event) { // Don't bother to schedule is there is already one out.
-              schedule_event = mutex->thread_holding->schedule_in(this, HRTIME_MSECONDS(10));
-            }
-            return 0;
-          }
-        }
+      Ptr<ProxyMutex> plugin_mutex;
 
-        this->api_current = this->api_current->next();
-        hook->invoke(eventmap[this->api_hookid], this);
+      if (hook->m_cont->mutex) {
+        plugin_mutex = hook->m_cont->mutex;
+        plugin_lock  = MUTEX_TAKE_TRY_LOCK(hook->m_cont->mutex, mutex->thread_holding);
```

Comment:
While you are in here, you may want to look at replacing MUTEX_TAKE_TRY_LOCK with MUTEX_TRY_LOCK.  I see that MUTEX_TAKE_TRY_LOCK has been deprecated.  But given the structure of this code with the possibility of a null mutex pointer, that may not be a very straightforward replacement.